packages/common/nbstore/src/storage/history.ts

```diff
@@ -0,0 +1,122 @@
+import { noop } from 'lodash-es';
+import {
+  applyUpdate,
+  Doc,
+  encodeStateAsUpdate,
+  encodeStateVector,
+  UndoManager,
+} from 'yjs';
+
+import { type DocRecord, DocStorage, type DocStorageOptions } from './doc';
+
+export interface HistoryFilter {
+  before?: Date;
+  limit?: Date;
+}
+
+export interface ListedHistory {
+  userId: string | null;
+  timestamp: Date;
+}
+
+export abstract class HistoricalDocStorage<
+  Options extends DocStorageOptions = DocStorageOptions,
+> extends DocStorage<Options> {
+  constructor(opts: Options) {
+    super(opts);
+
+    this.on('snapshot', snapshot => {
+      this.createHistory(snapshot.docId, snapshot).catch(noop);
+    });
+  }
+
+  override async setDocSnapshot(
+    snapshot: DocRecord,
+    prevSnapshot: DocRecord | null
+  ): Promise<boolean> {
+    const success = await this.setDocSnapshot(snapshot, prevSnapshot);
```

Comment:
super.xxx

Reply:
should be `this.upsertDocSnapshot `

Reply:
super.set is abstract function, delegate to another function is the only way to do this